opentracing/interceptor.go

```diff
@@ -0,0 +1,174 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package opentracing provides OpenTracing utilities.
+package opentracing
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/opentracing/opentracing-go"
+	"go.temporal.io/sdk/interceptor"
+)
+
+// TracerOptions are options provided to NewInterceptor or NewTracer.
+type TracerOptions struct {
+	// Tracer is the tracer to use. If not set, the global one is used.
+	Tracer opentracing.Tracer
+
+	// SpanContextKey is the context key used for internal span tracking (not to
+	// be confused with the context key OpenTracing uses internally). If not set,
+	// this defaults to an internal key (recommended).
+	SpanContextKey interface{}
+
+	// HeaderKey is the Temporal header field key used to serialize spans. If
+	// empty, this defaults to the one used by all SDKs (recommended).
+	HeaderKey string
+
+	// SpanStarter is a callback to create spans. If not set, this creates normal
+	// OpenTracing spans calling Tracer.StartSpan.
+	SpanStarter func(t opentracing.Tracer, operationName string, opts ...opentracing.StartSpanOption) opentracing.Span
+}
+
+type spanContextKey struct{}
+
+const defaultHeaderKey = "_tracer-data"
```

Comment:
Not sure if this can still be changed (backward compatibility?), but if we still could, consider adding "temporal" in the string?

Reply:
I have chosen not to change it but instead allow it to be customized. We also want this to work across SDKs, and Java uses this same string.